File: GFG/Stacks/infixtoprefix.cpp
#include<iostream>
#include<string>
#include"../ds.h"
using namespace std;

bool isOperator(char c) {
    return (c == '+' || c == '-' || c == '*' || c == '/');
}

// Function to get the precedence of an operator
int getPrecedence(char c) {
    if (c == '*' || c == '/')
        return 2;
    else if (c == '+' || c == '-')
        return 1;
    else
        return 0;
}

string infixtoprefix(string expr){
    int n=expr.size();
    Stack<char> st;
    int i=n-1;
    string res="";
    while(i>=0){
        if(isOperator(expr[i])){
            if(st.isEmpty()==false && getPrecedence(st.peek())>=getPrecedence(expr[i])){
                char t=st.pop();
                st.push(expr[i]);
                res=t+res;
            }
            else{
            st.push(expr[i]);
            }
        }
        else if(isalnum(expr[i])){
            res=expr[i]+res;
        }
        else if(expr[i]==')'){st.push(expr[i]);}
        else if(expr[i]=='('){
            while(st.isEmpty()==false && st.peek()!=')'){
                res=st.peek()+res;
                st.pop();
            }
            st.pop();
        }
        i--;
    }
    while(st.isEmpty()==false){
        if(st.peek()!=')'){
            res=st.peek()+res;
        }
        st.pop();
    }
    return res;
}

int main(){
    string eval="A * B + C * (D + E) - F / (G + H)";
    string res=infixtoprefix(eval);
    cout<<res<<endl;
    return 0;
}

File: GFG/Stacks/infixtoprefix.py
import ds

def is_operator(c):
    return c in ['+', '-', '*', '/']

def get_precedence(c):
    if c in ['*', '/']:
        return 2
    elif c in ['+', '-']:
        return 1
    else:
        return 0
    
def infixtoprefix(expr):
    n=len(expr)
    st=ds.Stack()
    res=""
    i=n-1
    while(i>=0):
        if(is_operator(expr[i])):
            if(st.isEmpty()==False and get_precedence(st.top())>get_precedence(expr[i])):
                tp=st.pop()
                st.push(expr[i])
                res=tp+res
            else:
                st.push(expr[i])
        elif(expr[i].isalnum()):
            res=expr[i]+res
        elif(expr[i]==')'):
            st.push(expr[i])
        elif(expr[i]=='('):
            while(st.isEmpty()==False and st.top()!=')'):
                res=st.top()+res
                st.pop()
            st.pop()
        i-=1
    while(st.isEmpty()==False):
        if(st.top() not in ['(',')']):
            res=st.top()+res
        st.pop()
        
    return res

eval="A * B + C * (D + E) - F / (G + H)"
res=infixtoprefix(eval)
print(res)